docs/design/validation.md

```diff
@@ -0,0 +1,2048 @@
+
+    for _, v := range vs {
+        result = append(result, v.Description(ctx))
+    }
+    return result
+}
+```
+
+To support passing through the provider instance, a separate interface type could be introduced that includes a function call with the `tfsdk.Provider` interface type:
+
+```go
+type ValueValidatorWithProvider interface {
+    ValueValidator
+    ValidateWithProvider(context.Context, provider tfsdk.Provider, path *tftypes.AttributePath, value attr.Value) error
+}
+```
+
+#### Attribute Value Validation Function Value Parameter
+
+Regardless the choice of concrete or interface types for the value validation functions, the parameters and returns for the implementations will play a crucial role on the extensibility and development experience.
+
+##### `attr.Value` Type
+
+The simplest implementation in the framework that could occur in all function types or interfaces is directly supplying an `attr.Value` and requiring implementations to handle all type conversion:
+
+```go
+func (v someValidator) Validate(ctx context.Context, path *tftypes.AttributePath, rawValue attr.Value) error {
+    value, ok := rawValue.(types.String)
+    
+    if !ok {
+        return fmt.Errorf("%s with incorrect type: %T", path, rawValue)
+    }
+
+    // ... rest of logic ...
+```
+
+This proposal would strongly encourage all implementations to handle value type conversions since using a stronger type in the function signature would risk panics that the framework cannot prevent and the compiler cannot check. Any error handling here could become inconsistent across implementations. This type conversion logic feels like an unnecessary burden on implementors and could reduce the developer experience as this logic would always need to be repeated with little to no actual utility.
+
+##### `types.T` Type
+
+If using an `attr.ValueValidator` interface approach, multiple new Go interface types could be created that define extensible value validation functions with strong typing. For example:
+
+```go
+// ValueValidator describes common validation functionality
+type ValueValidator interface {
+    Description(context.Context) string
+    MarkdownDescription(context.Context) string
+}
+
+// StringValueValidator describes String value validation
+type StringValueValidator interface {
+    ValueValidator
+    Validate(context.Context, *tftypes.AttributePath, types.String) error
+}
+```
+
+Then, this framework can handle the appropriate type conversions and error handling:
+
+```go
+// Validate performs all validation functions.
+//
+// Each type performs conversion or returns a conversion error
+// prior to executing the typed validation function.
+func (vs ValueValidators) Validate(ctx context.Context, path *tftypes.AttributePath, rawValue attr.Value) error {
+    for _, validator := range vs {
+        switch typedValidator := validator.(type) {
+        case StringValueValidator:
+            value, ok := rawValue.(types.String)
```

Comment:
We haven't talked together about the design of custom attributes, but implicitly all values can (should?) boil down to their complex and primitive value types. Otherwise that custom attribute design could have to recommend switching value types to an interface, which seems outside this design to discuss or postulate on as it would have broader effect.

Reply:
It seems like a type switch isn't the correct implementation detail in this scenario beyond what value types the framework implements -- custom value types would need to use a generic `attr.Value` (still provided as an escape hatch) and proposed `ValueAs` handling. This would really just serve to simplify framework validation helpers (or developers using `types.String` (for example).

Reply:
I think where we've landed on this is that, not as part of these proposals but potentially in the future, each type implementation (such as `types.String`, but others too!) could optionally offer helpers to simplify validation implementations to be strongly typed. The design in its current form does not prevent them and the details have been removed from the recommendations, instead left as a note within future considerations.